src/WinRT.Runtime/Marshalers.cs

```diff
@@ -485,158 +485,345 @@ class MarshalGeneric<T>
+                FromManaged = (Func<T, object>)(object)new Func<DateTimeOffset, object>(ABI.System.NonBlittableMarshallingStubs.DateTimeOffset_FromManaged);
+                CopyManaged = (Action<T, IntPtr>)(object)new Action<DateTimeOffset, IntPtr>(ABI.System.DateTimeOffset.CopyManaged);
+                DisposeMarshaler = ABI.System.NonBlittableMarshallingStubs.NoOpFunc;
+                DisposeAbi = ABI.System.NonBlittableMarshallingStubs.NoOpFunc;
+            }
+            else if (typeof(T).IsValueType)
+            {
+                // Value types can have custom marshaller types and use value types in places where we can't construct
+                // delegates in the same efficient way as with reference types. Use the fallback path in this case
+                HelperType = typeof(T).GetHelperType();
+                AbiType = typeof(T).GetAbiType();
+                MarshalerType = typeof(T).GetMarshalerType();
+                MarshalByObjectReferenceValueSupported = typeof(T).GetMarshaler2Type() == typeof(ObjectReferenceValue);
+
+                MarshalGenericFallback<T> fallback = new(HelperType);
+
+                CreateMarshaler = fallback.CreateMarshaler;
+                CreateMarshaler2 = MarshalByObjectReferenceValueSupported ? fallback.CreateMarshaler2 : CreateMarshaler;
+                GetAbi = fallback.GetAbi;
+                CopyAbi = fallback.CopyAbi;
+                FromAbi = fallback.FromAbi;
+                FromManaged = fallback.FromManaged;
+                CopyManaged = fallback.CopyManaged;
+                DisposeMarshaler = fallback.DisposeMarshaler;
+                DisposeAbi = fallback.DisposeAbi;
+                CreateMarshalerArray = fallback.CreateMarshalerArray;
+                GetAbiArray = fallback.GetAbiArray;
+                FromAbiArray = fallback.FromAbiArray;
+                FromManagedArray = fallback.FromManagedArray;
+                DisposeMarshalerArray = fallback.DisposeMarshalerArray;
+                DisposeAbiArray = fallback.DisposeAbiArray;
+            }
+            else
+            {
+                // Fallback case for all other types (could be anything, really). These would be reference types,
+                // which means we can make some assumptions on the shape of the helper type methods. Specifically,
+                // we expect the returned marshallers to be IObjectReference-s, and the ABI to be an IntPtr value.
+                HelperType = typeof(T).GetHelperType();
+                AbiType = typeof(T).GetAbiType();
+                MarshalerType = typeof(T).GetMarshalerType();
+                MarshalByObjectReferenceValueSupported = typeof(T).GetMarshaler2Type() == typeof(ObjectReferenceValue);
+#if NET
+                CreateMarshaler = HelperType.GetMethod("CreateMarshaler", BindingFlags.Public | BindingFlags.Static)?.CreateDelegate<Func<T, IObjectReference>>();

```

Comment:
As part of one of the past performance improvements, we made a bunch of these `lazy` because we saw doing a bunch of `GetMethod` calls adds to the cost especially when we weren't using all of them either.  Now with this change, we are going back to doing a bunch of `GetMethod` calls at the same time.  We shouldn't go back to `lazy` as it has its own concerns, but do you think we should consider doing it lazily without using lazy?  Or would doing that add much more to the size compared to the perf benefits?

Reply:
The problem is that all of those members are public fields, so we cannot really do any lazy initialization when accessing them. We'd have to assign them to some stub delegate with that additional logic (which is what we were doing before), but that would mean introducing a whole lot of additional generic delegates again, which would increase the binary size. I'd be inclined to say this is fine, considering two things:
- We now have a whole lot of hardcoded paths (eg. for all primitive types and for well known projected types) that skip this reflection logic entirely. All of those were being instantiated previously, so I'm thinking the fact that will no longer be the case should already help with performance on that front.
- `GetMethod()` and `CreateDelegate` should be much faster on NativeAOT compared to CoreCLR, so especially there it shouldn't really be necessary to pay the increased binary size to account for that.

What do you think about taking this change and then just potentially reevaluate the lazy initialization logic in the future in case we do get some new traces showing that being a problem (which might not actually be the case)? 🙂

Reply:
So, we have indeed reduced the scenarios where we use the generic marshalers as we source generate more of the code and there we know the actual type being used.  Given this brings size improvements and we have reduced the number of scenarios using this, lets go with this change and revisit if we notice perf issues as a result.

Reply:
Sounds good, thank you! 🙂